File: solutions/clojure/squeaky-clean/1/src/squeaky_clean.clj
(ns squeaky-clean
  (:require [clojure.string :as st]))

(defn clean
  "Replaces spaces with underscores
  Replaces control characters its code is in the range '\u0000' through '\u001F' or in the range '\u007F' through '\u009F' with 'CTRL'
  Omits characters that are not letters
  Replaces kebab with camel case 'à-ḃç' -> 'àḂç'
  Omits lowercase greek letters 
  "
  [s]
  (-> s
      (st/replace #"[\u0000-\u001F\u007F-\u009F]" "CTRL")
      (st/replace #"\-(.)" (comp st/upper-case second))
      (st/replace #" " "_")
      (st/replace #"[^_|^\p{L}]" "")
      (st/replace #"[α-ω]" "")))

File: solutions/csharp/squeaky-clean/1/SqueakyClean.cs
using System.Text.RegularExpressions;

public static class Identifier
{
    private const string Whitespace = @"\s";
    private const string ControlCharacter = @"\p{Cc}";
    private const string DashAndFollowingLetter = @"-(\p{L})";
    private const string NonLetterOrUnderscore = @"[^\p{L}_]";
    private const string GreekLowerCase = @"[α-ω]";

    public static string Clean(string identifier)
    {
        return identifier
                // Space with underscore
                .RxReplace(Whitespace, "_")
                // Control chars with "CTRL"
                .RxReplace(ControlCharacter, "CTRL")
                // Kebab with CamelCase
                .RxReplace(DashAndFollowingLetter, match => match.Groups[1].Value.ToUpperInvariant())
                // omit chars that are not letters except underscore
                .RxReplace(NonLetterOrUnderscore, "")
                // omit greek letters
                .RxReplace(GreekLowerCase, "")
            ;

    }
    
}

static class StringExtensions
{
    public static string RxReplace(this string input, string pattern, string replacement)
        => Regex.Replace(input, pattern, replacement);

    public static string RxReplace(this string input, string pattern, MatchEvaluator evaluator)
        => Regex.Replace(input, pattern, evaluator);
}
